src/Stratis.Bitcoin/P2P/PeerConnector.cs

```diff
@@ -0,0 +1,176 @@
+
+        /// <summary> Peer address manager instance, see <see cref="IPeerAddressManager"/>.</summary>
+        private readonly IPeerAddressManager peerAddressManager;
+
+        /// <summary> What peer types (by <see cref="PeerIntroductionType"/> this connector should find and connect to.</summary>
+        private readonly PeerIntroductionType peerIntroductionType;
+
+        internal RelatedPeerConnectors RelatedPeerConnector { get; set; }
+
+        /// <summary> Specification of requirements the <see cref="PeerConnector"/> has when connect to other peers.</summary>
+        internal readonly NodeRequirement Requirements;
+
+        internal PeerConnector(Network network,
+            INodeLifetime nodeLifeTime,
+            NodeConnectionParameters parameters,
+            NodeRequirement nodeRequirements,
+            Func<IPEndPoint, byte[]> groupSelector,
+            IAsyncLoopFactory asyncLoopFactory,
+            IPeerAddressManager peerAddressManager,
+            PeerIntroductionType peerIntroductionType)
+        {
+            this.asyncLoopFactory = asyncLoopFactory;
+            this.ConnectedPeers = new NodesCollection();
+            this.groupSelector = groupSelector;
+            this.MaximumNodeConnections = 8;
+            this.network = network;
+            this.nodeLifetime = nodeLifeTime;
+            this.ParentParameters = parameters;
+            this.peerAddressManager = peerAddressManager;
+            this.peerIntroductionType = peerIntroductionType;
+            this.Requirements = nodeRequirements;
+
+            this.currentParameters = this.ParentParameters.Clone();
+            this.currentParameters.TemplateBehaviors.Add(new PeerConnectorBehaviour(this));
+            this.currentParameters.ConnectCancellation = this.nodeLifetime.ApplicationStopping;
+        }
+
+        internal void AddNode(Node node)
+        {
+            Guard.NotNull(node, nameof(node));
+
+            this.ConnectedPeers.Add(node);
+        }
+
+        internal void StartConnectAsync()
+        {
+            this.asyncLoop = this.asyncLoopFactory.Run(nameof(this.ConnectAsync), async token =>
+            {
+                if (this.ConnectedPeers.Count < this.MaximumNodeConnections)
+                    await ConnectAsync();
+            },
+            this.nodeLifetime.ApplicationStopping,
+            repeatEvery: TimeSpans.Second);
+        }
+
+        private Task ConnectAsync()
+        {
+            Node node = null;
+
+            try
+            {
+                var peer = FindPeerToConnectTo();
+                if (peer == null)
+                    return Task.CompletedTask;
+
+                using (var timeoutTokenSource = CancellationTokenSource.CreateLinkedTokenSource(this.nodeLifetime.ApplicationStopping))
+                {
+                    timeoutTokenSource.CancelAfter(5000);
+
+                    this.peerAddressManager.PeerAttempted(peer.Endpoint, DateTimeProvider.Default.GetUtcNow());
```

Comment:
This should go after the handshake no?

Reply:
We need to set the last connection attempt on PeerAddress before we try and connect. If connection succeeds we set the LastConnectionSuccess and LastHandshaked properties from the event handlers.

Reply:
Ah right awesome.